doc/event-alarm-framework/events-producer.md

```diff
@@ -0,0 +1,212 @@
+
+
+![image](https://user-images.githubusercontent.com/47282725/156947460-66d08b3d-c981-4413-b0d5-232643dfba01.png)
+
+
+![image](https://user-images.githubusercontent.com/47282725/159061158-30ff3c8a-5fc0-4af2-8822-6bfc67b2329c.png)
+
+
+
+## Requirements
+### Events
+1. Events are defined with schema.
+2. Every event is identified by tag, which is unique within a process with zero or more event specific parameters.
+3. Events are static (*don't change*) across releases, but can be deprecated in newer releases.
+4. Every event is described in YANG schema.
+5. YANG schema files for all events are available in a single location for NB clients to refer in Switch.
+
+### Event detection
+1. The method of detection can be any.
+2. This can vary across events.
+3. Syslog messages might be used to detect an event or some custom queries or code-update to report events directly or ...
+4. There can be multiple event detectors running under different scopes, concurrently.
+
+### Event reporting
+1. Event detectors stream the events with structured data.
+2. The streaming supports one or more local listeners to receive streams from multiple detectors (many-to-many).
+3. The structured data is per YANG definition.
+4. RFE: A listener to update redis/any other persistence destination with current event status.
+
+### Event exporter
+1. Telemetry container receive all the events reported from all the event detectors.
+2. Telemetry container provides support for streaming out received events to multiple external clients.
+3. RFE: Telemetry container upon restart will use redis/any other persistence to get the latest on missed updates during its down time and stream out.
+
+### Event reliability
+There are two kinds of reliability.
+1. Events are not modified across releases (*except deprecation*). We may allow addition of new params, as long as they don't affect existing params.
+2. Events are verified to fire as expected in every image release. 
+3. Ensure the perf goals are met.
+
+#### Protection:
+
+1. The unit tests are required to ensure the immutability of event definition across releases. This would help block a PR that mutates a definition.
+2. The nightly tests are required to ensure the process does fire the event, when the scenario occurs and the o/p is per definition. This would help block the release.
+3. The unit tests & nightly tests are required for every event.
+4. A separate stress test is required to ensure the performance rate of 10K events/sec and 99.5% of reliability end-to-end.
+</br>
+
+## Design
+
+### YANG schema
+1. Schema defines the description of the event, its unique tag and all the possible parameters.
+2. Schema can be maintained in multiple files, preferably one per process/continer/host.
+3. All the schema files are copied into one single location (e.g. /usr/shared/sonic/events).
+4. The schema for processes running in host are copied into this location at image creation/install time.
```

Comment:
>creation/install [](http://example.com/codeflow?start=83&length=16)

When did them copied? compilation time or installation time?

Reply:
From you answer of next comment, seem you mean container first run time, not "image creation/install time".